### ISSUE TITLE
実装: bidding PPOに♠♥♦♣ 32件ずつのstratified minibatchを追加する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#317]
Related: #202, #303, #304, #305, #310
Base branch: `feature/issue-202-non-playing-ai`

## 背景

non-playing bidding PPOでは、suit collapseの調査を進めた結果、#304 のcounterfactual評価で、クラブ適性局面では強制♣選択のrewardが♠/♥/♦より同等以上になるケースが多い一方、学習済みActorは♣確率をほぼ0まで落としていることが確認された。

このため reward/game dynamicsだけではなく、on-policy PPO更新中のサンプリング偏り・softmax競合・rare actionの自己強化的消失が主要因の候補になっている。

一方、C++/CUDA rollout高速化により、2000 logical seeds / 10000 actual games のrolloutは約36秒まで短縮できており、学習条件のablationを回しやすくなった。

次の比較として、bidding PPOのmini-batchを、各sampleの手札から求めた「最も適性の高いsuit」で層化し、1 batch 128件を以下で固定する実験を行いたい。

```text
♠適性 32件
♥適性 32件
♦適性 32件
♣適性 32件
合計 128件
```

これは最終方式を決める実装ではなく、suit collapseがbalanced exposureで改善するかを見るablationである。

## 目的

既存のrandom minibatch方式を壊さず、bidding PPOに選択可能な stratified minibatch sampler を追加する。

新方式では1 mini-batchを必ず:

```text
Spades   32
Hearts   32
Diamonds 32
Clubs    32
```

で構成する。

PPO objective、reward、advantage、entropy、model architecture、rollout semanticsは変更しない。

## Suit適性の定義

各training sampleのbidding observationに対応する手札について、既存 `RuleBasedAgent.evaluateHandForTrump(hand, suit)` と同等の評価を4 suitへ適用し、最大scoreのsuitを `strongestSuit` とする。

tieは既存調査 #303/#304 と同一にする。

```text
S > H > D > C
```

独自の新しいheuristicを作らない。

可能なら既存ロジックを共通利用し、Python側で再実装する場合もgolden parity testを入れる。

### Batch composition

`batch_size=128` のとき:

```text
S pool -> 32
H pool -> 32
D pool -> 32
C pool -> 32
```

を取得し、128件をbatch内部でshuffleしてoptimizerへ渡す。

このIssueでは32/32/32/32固定のablationを第一対象とする。

### Pool不足 / oversampling

各suit poolはepoch開始時に独立shuffleする。

あるpoolを使い切った場合、そのpoolを再shuffleして先頭から再利用する。

したがって少数suitはoversamplingされる。

これは意図した実験条件であり、隠さずtraining diagnostics / provenanceへ記録する。

epochのbatch数は、全dataset sampleを少なくとも概ね1回処理できるよう、最大poolまたは総sample数基準で決める。具体方式は実装で選んでよいが、以下を満たすこと。

- batchは常に32/32/32/32
- 小さいpoolを理由に早期打ち切りして大量sampleを捨てない
- oversampling回数を計測できる
- seed固定でreproducible

### batch_size制約

stratified modeは今回 `batch_size=128` のみを正式サポートしてよい。

他batch sizeを許可する場合は4で割り切れ、各suit同数になることを明確にする。
曖昧な端数配分はしない。

## CLI / config

既存random batchingをdefaultとして維持する。

例:

```text
--bidding-minibatch-strategy random
--bidding-minibatch-strategy strongest-suit-balanced
```

命名は既存CLI規則へ合わせてよい。

stratified modeではbatch compositionをprovenanceへ保存する。

最低限:

- minibatch strategy
- strongest suit definition/version
- per-suit target count
- epochごとのoriginal pool size
- epochごとのdraw count / oversampling count

resume時にstrategyが違う場合はfail-closeする。

legacy run configにこのfieldがない場合は `random` と解釈し、既存resume compatibilityを壊さない。

## 学習objectiveは変更しない

今回変更しないもの:

- PPO clipped objective
- behavior log probability
- advantage計算
- advantage normalization設定
- entropy coefficient
- learning rate
- epochs
- reward v3
- relative terminal reward
- Actor/Critic architecture

stratified samplingだけの効果を切り分ける。

## Diagnostics

最低限iteration summaryまたはtraining reportで以下を確認できるようにする。

```text
original strongestSuit counts:
  S / H / D / C

sampled counts used for optimizer:
  S / H / D / C

oversampled counts:
  S / H / D / C

number of minibatches
```

stratified modeでは optimizer入力が32/32/32/32であることをunit testでも保証する。

## Tests

最低限:

1. strongestSuit classification parity
2. 4 poolが十分ある場合、各batchが32/32/32/32
3. C pool等が小さい場合、reshuffle + reuseされても32/32/32/32を維持
4. small pool reuseがdeterministic seedで再現可能
5. random modeの既存挙動を壊さない
6. stratified modeでoptimizer stepが完走する
7. checkpoint/resumeでstrategy provenance維持
8. legacy config missing strategy -> randomとしてresume可能
9. random <-> stratified mismatch resumeはfail-close
10. CPU/CUDA training device双方でsampler semanticsが一致

## 最初の比較実験

実装後、まず以下を推奨する。

```text
model hidden: 512,512,256,256
entropy: 0.1
learning rate: 1e-4
advantage normalization: OFF
epochs: 4
batch size: 128
minibatch: strongest-suit-balanced 32/32/32/32
simulation backend: cpp
inference: cuda
training: cuda
games / iteration: 2000 logical seeds
```

random minibatch baselineと、他条件を同一にして比較する。

最低限比較:

- pass rate
- Napoleon rate
- declaration success
- mean learning reward
- selected suit counts
- strongestSuit=C subsetの P(C)
- RuleBased=C subsetの P(C)
- S/H/D/C probability distribution
- evaluation candidate vs RuleBased

## Acceptance Criteria

- bidding PPOでrandom / strongest-suit-balancedを切替可能
- balanced modeの各mini-batchが32/32/32/32
- 少数poolはreshuffle/reuseされ、早期打ち切りしない
- oversampling量がdiagnosticsへ出る
- strongestSuit定義が#303/#304と一致
- existing PPO objective / reward / advantage semantics不変
- random mode regressionなし
- checkpoint/resume/provenance対応
- CUDA training smoke完走
- 1 iteration C++/CUDA rollout -> stratified PPO update -> exportまで完走

## 非目標

- reward shaping
- suit-specific reward
- counterfactual auxiliary loss
- BC / imitation loss
- KL regularization
- suit-family entropy
- policy architecture変更
- rollout/game semantics変更
- balanced samplerを最終方式と決めること

## Git運用

作業場所:
`~/napoleon_ws/napoleon-web-issue-202`

`feature/issue-202-non-playing-ai` からIssue専用ブランチを作成する。
PR baseは必ず `feature/issue-202-non-playing-ai`。
mainにはmergeしない。

#303/#304 diagnostics/scripts等の未追跡ファイルを誤ってcommitしないこと。

Codex review最大4回、review後修正最大3回。
blocking問題がなくなればReady for review。
テスト・review通過後は親ブランチへmergeする。
